app/src/game-engine/local-engine/GameState.ts

```diff
@@ -11,21 +11,24 @@ export class GameState {
   public capturedPieces: Array<Piece>;
   public onCheck: boolean;
   public result: GameResult;
+  public castlingFlags: boolean[][];
 
   constructor(
     board: Board,
     currentTurn: PieceColor,
     history: Array<Move>,
     capturedPieces: Array<Piece>,
     onCheck: boolean,
-    result: GameResult
+    result: GameResult,
+    castlingFlags: boolean[][]
```

Comment:
I like this idea. What about replacing this with a more general key-value dictionary? Say:
```typescript
// This dictionary might be placed in a more global file.
interface Dictionary<T> {
    [Key: string]: T;
}

export class GameState {
    board: Board,
    currentTurn: PieceColor,
	// other members...
	flags: Dictionary<string>, // Or Dictionary<any>.
}
```
Then, each rule could set or check any flag it needs without adding new members to `GameState`. In the case of castling, you could implement something like:
```typescript
if (!state.flags["white-castled"]) {
	// Castling logic...
    newState.flags["white-castled"] = true;
}
```